File: pc/protocol.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"io"
)

const (
	protoVersion = "0.1.0"
	maxHeaderLen = 1 << 20 // 1 MiB
	maxPayload   = 1 << 30 // 1 GiB sanity cap
)

// Header is the JSON header carried before every payload.
type Header struct {
	Type       string `json:"type"`
	PayloadLen int64  `json:"payloadLen"`
	SessionID  string `json:"sessionId,omitempty"`
	FileID     string `json:"fileId,omitempty"`
	FileName   string `json:"fileName,omitempty"`
	FileSize   int64  `json:"fileSize,omitempty"`
	ChunkIndex int64  `json:"chunkIndex,omitempty"`
	TotalBytes int64  `json:"totalBytes,omitempty"`
	DeviceName string `json:"deviceName,omitempty"`
	DeviceType string `json:"deviceType,omitempty"`
	Version    string `json:"version,omitempty"`
	FileIndex  int    `json:"fileIndex,omitempty"`
	FileCount  int    `json:"fileCount,omitempty"`
	Message    string `json:"message,omitempty"`
	Port       int    `json:"port,omitempty"`
	IP         string `json:"ip,omitempty"`
	Text       string `json:"text,omitempty"`
	RelPath    string `json:"relPath,omitempty"`
	MD5        string `json:"md5,omitempty"`    // file_end 时整文件 MD5（32 位小写 hex）
	Mtime      int64  `json:"mtime,omitempty"`  // file_meta 时源文件修改时间（Unix 秒）
	Sync       bool   `json:"sync,omitempty"`   // file_meta 时为目录同步传输（落到同步专用目录）
	Offset     int64  `json:"offset,omitempty"` // 已废弃（断点续传移除），仅向后兼容
}

func helloHeader(name, dtype string) Header {
	return Header{Type: "hello", DeviceName: name, DeviceType: dtype, Version: protoVersion, Port: selfListenPort, IP: primaryIP()}
}

// writeFrame writes a length-prefixed header + payload.
func writeFrame(w io.Writer, h Header, payload []byte) error {
	h.PayloadLen = int64(len(payload))
	hb, err := json.Marshal(h)
	if err != nil {
		return err
	}
	if len(hb) > maxHeaderLen {
		return errors.New("header too large")
	}
	var lenbuf [4]byte
	binary.BigEndian.PutUint32(lenbuf[:], uint32(len(hb)))
	if _, err := w.Write(lenbuf[:]); err != nil {
		return err
	}
	if _, err := w.Write(hb); err != nil {
		return err
	}
	if len(payload) > 0 {
		if _, err := w.Write(payload); err != nil {
			return err
		}
	}
	return nil
}

// readFrame reads a full frame. The returned payload is only valid for the
// duration of the caller's processing (a fresh slice is allocated each call).
func readFrame(r io.Reader) (Header, []byte, error) {
	var lenbuf [4]byte
	if _, err := io.ReadFull(r, lenbuf[:]); err != nil {
		return Header{}, nil, err
	}
	hlen := binary.BigEndian.Uint32(lenbuf[:])
	if hlen > maxHeaderLen {
		return Header{}, nil, errors.New("header too large")
	}
	hb := make([]byte, hlen)
	if _, err := io.ReadFull(r, hb); err != nil {
		return Header{}, nil, err
	}
	var h Header
	if err := json.Unmarshal(hb, &h); err != nil {
		return Header{}, nil, err
	}
	var payload []byte
	if h.PayloadLen > 0 {
		if h.PayloadLen > maxPayload {
			return Header{}, nil, errors.New("payload too large")
		}
		payload = make([]byte, h.PayloadLen)
		if _, err := io.ReadFull(r, payload); err != nil {
			return Header{}, nil, err
		}
	}
	return h, payload, nil
}

File: android/app/src/main/java/com/lantransfer/app/Protocol.kt
package com.lantransfer.app

import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.InputStream
import java.io.OutputStream

object Protocol {
    const val VERSION = "0.1.0"

    class Header {
        var type: String = ""
        var payloadLen: Long = 0
        var sessionId: String? = null
        var fileId: String? = null
        var fileName: String? = null
        var fileSize: Long = 0
        var chunkIndex: Long = 0
        var totalBytes: Long = 0
        var deviceName: String? = null
        var deviceType: String? = null
        var version: String? = null
        var fileIndex: Int = 0
        var fileCount: Int = 0
        var message: String? = null
        var port: Int = 0
        var ip: String? = null
        var text: String? = null
        var relPath: String? = null
        var md5: String? = null      // file_end 时整文件 MD5（32 位小写 hex）
        var mtime: Long = 0          // file_meta 时源文件修改时间（Unix 秒）
        var sync: Boolean = false    // file_meta 时为目录同步传输（落到电脑同步专用目录）
        var offset: Long = 0         // 已废弃（断点续传移除），仅向后兼容

        fun toJson(): JSONObject {
            val o = JSONObject()
            o.put("type", type)
            o.put("payloadLen", payloadLen)
            sessionId?.let { o.put("sessionId", it) }
            fileId?.let { o.put("fileId", it) }
            fileName?.let { o.put("fileName", it) }
            if (fileSize != 0L) o.put("fileSize", fileSize)
            if (chunkIndex != 0L) o.put("chunkIndex", chunkIndex)
            if (totalBytes != 0L) o.put("totalBytes", totalBytes)
            deviceName?.let { o.put("deviceName", it) }
            deviceType?.let { o.put("deviceType", it) }
            version?.let { o.put("version", it) }
            if (fileIndex != 0) o.put("fileIndex", fileIndex)
            if (fileCount != 0) o.put("fileCount", fileCount)
            message?.let { o.put("message", it) }
            if (port != 0) o.put("port", port)
            ip?.let { o.put("ip", it) }
            text?.let { o.put("text", it) }
            relPath?.let { o.put("relPath", it) }
            md5?.let { o.put("md5", it) }
            if (mtime != 0L) o.put("mtime", mtime)
            if (sync) o.put("sync", true)
            if (offset != 0L) o.put("offset", offset)
            return o
        }

        companion object {
            fun fromJson(o: JSONObject): Header {
                val h = Header()
                h.type = o.optString("type")
                h.payloadLen = o.optLong("payloadLen", 0)
                h.sessionId = o.optString("sessionId").takeIf { it.isNotEmpty() }
                h.fileId = o.optString("fileId").takeIf { it.isNotEmpty() }
                h.fileName = o.optString("fileName").takeIf { it.isNotEmpty() }
                h.fileSize = o.optLong("fileSize", 0)
                h.chunkIndex = o.optLong("chunkIndex", 0)
                h.totalBytes = o.optLong("totalBytes", 0)
                h.deviceName = o.optString("deviceName").takeIf { it.isNotEmpty() }
                h.deviceType = o.optString("deviceType").takeIf { it.isNotEmpty() }
                h.version = o.optString("version").takeIf { it.isNotEmpty() }
                h.fileIndex = o.optInt("fileIndex", 0)
                h.fileCount = o.optInt("fileCount", 0)
                h.message = o.optString("message").takeIf { it.isNotEmpty() }
                h.port = o.optInt("port", 0)
                h.ip = o.optString("ip").takeIf { it.isNotEmpty() }
                h.text = o.optString("text").takeIf { it.isNotEmpty() }
                h.relPath = o.optString("relPath").takeIf { it.isNotEmpty() }
                h.md5 = o.optString("md5").takeIf { it.isNotEmpty() }
                h.mtime = o.optLong("mtime", 0)
                h.sync = o.optBoolean("sync", false)
                h.offset = o.optLong("offset", 0)
                return h
            }
        }
    }

    fun hello(name: String, type: String): Header {
        val h = Header()
        h.type = "hello"
        h.deviceName = name
        h.deviceType = type
        h.version = VERSION
        h.port = DEFAULT_PORT
        h.ip = Discovery.primaryIP()
        return h
    }

    fun write(out: OutputStream, h: Header, payload: ByteArray? = null) {
        val p = payload ?: ByteArray(0)
        h.payloadLen = p.size.toLong()
        val json = h.toJson().toString().toByteArray(Charsets.UTF_8)
        val dout = DataOutputStream(out)
        dout.writeInt(json.size)
        dout.write(json)
        if (p.isNotEmpty()) dout.write(p)
        dout.flush()
    }

    fun read(input: InputStream): Pair<Header, ByteArray?>? {
        return try {
            val din = DataInputStream(input)
            val len = din.readInt()
            if (len <= 0 || len > (1 shl 20)) return null
            val jsonBytes = ByteArray(len)
            din.readFully(jsonBytes)
            val h = Header.fromJson(JSONObject(String(jsonBytes, Charsets.UTF_8)))
            var payload: ByteArray? = null
            if (h.payloadLen > 0) {
                if (h.payloadLen > (1L shl 30)) return null
                payload = ByteArray(h.payloadLen.toInt())
                din.readFully(payload)
            }
            h to payload
        } catch (e: Exception) {
            null
        }
    }
}
